prov/verbs/src/verbs_cq.c

```diff
@@ -230,47 +230,169 @@ static void fi_ibv_cq_read_data_entry(struct ibv_wc *wc, int i, void *buf)
-		return ret;
-
-	if (wc->opcode == IBV_WC_RECV || wc->opcode == IBV_WC_RECV_RDMA_WITH_IMM)
-		return ret;
-
-	/* TODO Handle the case when app posts a send with same wr_id */
-	if ((wc->wr_id & cq->wr_id_mask) != cq->send_signal_wr_id)
-		return ret;
-
-	entry = slist_remove_first_match(&cq->ep_list, fi_ibv_match_ep_id, (void *)wc->wr_id);
+	entry = slist_remove_first_match(&cq->ep_list,
+					 fi_ibv_match_ep_id,
+					 (void *)wc->wr_id);
 	if (!entry) {
 		VERBS_WARN(FI_LOG_CQ, "No matching EP for :"
 			   "given signaled send completion\n");
 		return -FI_EOTHER;
 	}
 	epe = container_of(entry, struct fi_ibv_msg_epe, entry);
 	ofi_atomic_sub32(&epe->ep->unsignaled_send_cnt,
-			VERBS_SEND_SIGNAL_THRESH(epe->ep));
+			 VERBS_SEND_SIGNAL_THRESH(epe->ep));
 	ofi_atomic_dec32(&epe->ep->comp_pending);
 	util_buf_release(cq->epe_pool, epe);
 
-	return 0;
+	return FI_SUCCESS;
+}
+
+static inline int fi_ibv_poll_outstanding_cq(struct fi_ibv_msg_ep *ep,
+					     struct fi_ibv_cq *cq)
+{
+	struct fi_ibv_wre *wre;
+	struct fi_ibv_wce *wce;
+	struct fi_ibv_msg_ep *tmp_ep;
+	struct ibv_wc wc;
+	ssize_t ret;
+
+	ret = ibv_poll_cq(cq->cq, 1, &wc);
+	if (ret <= 0)
+		return ret;
+
+	if ((wc.opcode == IBV_WC_RECV) ||
+	    (wc.opcode == IBV_WC_RECV_RDMA_WITH_IMM) ||
+	    ((wc.wr_id & cq->wr_id_mask) != cq->send_signal_wr_id)) {
+		wre = (struct fi_ibv_wre *)(uintptr_t)wc.wr_id;
+		if (wre->ep) {
+			if ((wre->ep != ep) &&
+			    (wc.status != IBV_WC_WR_FLUSH_ERR)) {
+				wce = util_buf_alloc(cq->wce_pool);
+				if (!wce)
+					return -FI_EAGAIN;
+				memset(wce, 0, sizeof(*wce));
+				wc.wr_id = (uintptr_t)wre->context;
+				wce->wc = wc;
+				slist_insert_tail(&wce->entry, &cq->wcq);
+			}
+		} else {
+			/* if `ep == NULL` it means that WRE
+			 * has already been released */
+			return ret;
+		}
+		tmp_ep = wre->ep;
+		wre->ep = NULL;
+		util_buf_release(tmp_ep->wre_pool, wre);
+	} else {
+		ret = fi_ibv_handle_internal_signal_wc(cq, &wc);
+		return ret ? ret : 1;
+	}
+
+	return ret;
+}
+
+void fi_ibv_cleanup_cq(struct fi_ibv_msg_ep *ep)
+{
+	int ret;
+	struct fi_ibv_wre *wre;
+	struct dlist_entry *tmp;
+
+	fastlock_acquire(&ep->rcq->lock);
+	do {
+		ret = fi_ibv_poll_outstanding_cq(ep, ep->rcq);
+	} while (ret > 0);
+	/* Handle WRs for which there were no appropriate WCs */
+	dlist_foreach_container_safe(&ep->wre_list, struct fi_ibv_wre,
+				     wre, entry, tmp) {
+		if ((ep == wre->ep) &&
+		    (wre->wr.type == IBV_RECV_WR)) {
+			dlist_remove(&wre->entry);
+			wre->ep = NULL;
+			util_buf_release(ep->wre_pool, wre);
+		}
+	}
+	fastlock_release(&ep->rcq->lock);
+
+	fastlock_acquire(&ep->scq->lock);
+	do {
+		ret = fi_ibv_poll_outstanding_cq(ep, ep->scq);
+	} while (ret > 0);
+	dlist_foreach_container_safe(&ep->wre_list, struct fi_ibv_wre,
+				     wre, entry, tmp) {
+		if ((ep == wre->ep) &&
+		    (wre->wr.type == IBV_SEND_WR)) {
+			dlist_remove(&wre->entry);
+			wre->ep = NULL;
+			util_buf_release(ep->wre_pool, wre);
+		}
+	}
```

Comment:
I'm afraid that it will lead to performance degradation on verbs level, because we should serialize access to `ibv_poll_cq` anyway (it could be done only by locking of CQ->lock).
We need to held ep->lock and cq->lock, If I got your proposal correctly and considering my remark below. But it can help us only in `fi_ibv_cleanup_cq` where we go though wre_list.

Reply:
Yes, avoiding holding both locks is preferred. I believe that's doable but I think it would require significant refactoring of the poll_cq and cq read functions. May be we can defer it for later.